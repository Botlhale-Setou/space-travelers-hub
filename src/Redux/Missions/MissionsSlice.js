/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  missions: [],
  status: 'idle',
  error: null,
};

const fetchURL = 'https://api.spacexdata.com/v3/missions';

export const fetchMissionData = createAsyncThunk('spacetravelers/missions/fetch_data', async () => {
  const response = await axios.get(fetchURL);
  if (response.data) {
    return response.data;
  }
  return [];
});

const reArrangeMission = (missions) => {
  const newMissions = [];
  Object.entries(missions).forEach((mission) => {
    newMissions.push({
      id: mission[1].mission_id,
      name: mission[1].mission_name,
      description: mission[1].description,
      reserved: false,
    });
  });
  return newMissions;
};

const missionsSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    updateMission(state, action) {
      return {
        ...state,
        missions: state.missions.map((mission) => (mission.id === action.payload
          ? { ...mission, reserved: !mission.reserved }
          : mission)),
      };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMissionData.pending, (state) => {
      state.status = 'loading';
    })
      .addCase(fetchMissionData.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.missions = reArrangeMission(action.payload);
      })
      .addCase(fetchMissionData.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message;
      });
  },
});

export const { updateMission } = missionsSlice.actions;
export default missionsSlice;
