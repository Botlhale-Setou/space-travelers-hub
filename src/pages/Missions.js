import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Table } from 'react-bootstrap';
import Button from 'react-bootstrap/Button';
import { fetchMissionData, updateMission } from '../Redux/Missions/MissionsSlice';

function Missions() {
  const dispatch = useDispatch();
  const { missions, status, error } = useSelector((state) => state.missions);

  useEffect(() => {
    if (missions.length === 0) {
      dispatch(fetchMissionData());
    }
  }, [missions, dispatch]);

  if (status === 'loading') {
    return <div>Loading...</div>;
  }

  if (status === 'failed') {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <Table striped bordered hover size="sm" className="tableMission">
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th className="enlargeCol">
            Status
          </th>
          <th className="enlargeCol">
            &nbsp;
          </th>
        </tr>
      </thead>
      <tbody>
        {missions.map((item) => (
          <tr key={item.id}>
            <td>{item.name}</td>
            <td>{item.description}</td>
            <td className="text-middle text-center">
              <badge
                className={item.reserved ? 'yReserve' : 'nReserve'}
              >
                {item.reserved ? 'Active Member' : 'Not A Member'}
              </badge>
            </td>
            <td className="text-middle text-center">
              <Button
                variant={item.reserved ? 'outline-danger' : 'outline-primary'}
                type="button"
                className={item.reserved ? 'yReserve2' : 'nReserve2'}
                onClick={() => { dispatch(updateMission(item.id)); }}
              >
                {item.reserved ? 'Leave Mission' : 'Join Mission'}
              </Button>
            </td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
}

export default Missions;
