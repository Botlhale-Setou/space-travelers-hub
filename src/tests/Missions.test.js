import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../Redux/ConfigureStore';
import '@testing-library/jest-dom';
import Missions from '../pages/Missions';
import MissionsSlice, { updateMission } from '../Redux/Missions/MissionsSlice';

describe('Missions page', () => {
  test('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('updateMission', () => {
    it('should toggle the reserved property of the mission with the specified id', () => {
      const state = {
        missions: [
          {
            id: '1', name: 'MissionA', description: 'A mission', reserved: false,
          },
          {
            id: '2', name: 'MissionB', description: 'B mission', reserved: false,
          },
          {
            id: '3', name: 'MissionC', description: 'C mission', reserved: false,
          },
        ],
        status: 'succeeded',
        error: null,
      };
      const action = {
        type: updateMission.type,
        payload: '2',
      };
      const expectedState = {
        missions: [
          {
            id: '1', name: 'MissionA', description: 'A mission', reserved: false,
          },
          {
            id: '2', name: 'MissionB', description: 'B mission', reserved: true,
          },
          {
            id: '3', name: 'MissionC', description: 'C mission', reserved: false,
          },
        ],
        status: 'succeeded',
        error: null,
      };
      const result = MissionsSlice.reducer(state, action);
      expect(result).toEqual(expectedState);
    });
  });
});
