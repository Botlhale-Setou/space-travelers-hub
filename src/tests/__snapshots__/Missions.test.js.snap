// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions page renders correctly 1`] = `
<table
  className="tableMission table table-sm table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th
        className="enlargeCol"
      >
        Status
      </th>
      <th
        className="enlargeCol"
      >
        Â 
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
